Tls injection mechanism

sergejskozlovics opened the pull request:
I have developed a TLS injection mechanism. It can be used to inject PQC algorithms (KEMs and signature algorithms) into TLS 1.3. The injection mechanism is located in the org.bouncycastle.tls.injection package, where the InjectionPoint class has been defined. Then, there are many places where InjectionPoint is consulted for injected KEMs and signature algorithms, which are added to the corresponding lists of supported TLS algorithms.

peterdettman commented:
Hi Sergejs, thanks for the PR. It will be a while before a full review can be done, but on a first pass I have seen some basic issues:

- core project cannot depend on tls project (only the reverse). So I guess the basic injection concept would have to live in core and be used/built-upon by tls.
- tls project supports still minimum java version 5 (actually we still build also for jdk 1.4 with org.bouncycastle.tls classes only - no JSSE). I see at least some usage of Java 8 stream API, which should be replaced, and perhaps there are others. There are src/main/jdk??? folders to provide multi-release override classes if needed.
- There's some incidental reformatting that should be removed from the patch; I see various imports collapsed to ".*", braces moved, some switch case indentation, and there's probably more.

We are certainly interested in getting PQC algorithms into TLS, but our intention would be to proceed by publishing algorithms in the BC provider as usual, then having TLS add the registrar entries from the relevant standards and support them when it can find the algorithm via the provider. Perhaps you could comment a little about what you have in mind in terms of the basic "injection" idea here, particularly in comparison to the existing provider-based mechanism.

yuhh0328 committed:
Updates : add tls-kyber

Commit:
TODO: cover tests for EDDSA_LEGACY in BcImplProvider.

Commit:
TODO: cover tests for EDDSA_LEGACY in BcImplProvider.

Commit:
changes of build.gradle

Commit:
Merge branch 'main' into pg-jacoco

Commit:
Fix :pg:checkstyleMain.

Commit:
Add tests for BcImplProvider, ECDHPublicBCPGKey, ECSecretBCPGKey etc.

Commit:
Add a system property enableCamelliaKeyWrapping for camellia key wrapping

yuhh0328 committed:
Merge branch 'bcgit:main' into tls-kyber

Commit:
1. rename (PQC -> KEM)
2. add TlsCryptoParameters

Commit:
remove TlsKemMode

Commit:
Add tests to PGPOnePassSignatureList, PGPOnePassSignature, PGPEncryptedDataList, PGPCompressedDataGenerator and PGPUtil.
Extract PGPDefaultSignatureGenerator.getAttriubtesHash from PGPSignature and PGPSignatureGenerator.Invert PGPSignatureGenerator.packetPresent to packetNotPresent.

Commit:
Add tests to PGPUtils

Commit:
setWithAEAD in JcePGPDataEncryptorBuilder and BcPGPDataEncryptorBuilder allow Camellia. Add tests to BcImplProvider, BcKeyFingerprintCalculator, BcPGPContentVerifierBuilderProvider, BcPGPDataEncryptorBuilder and JcePGPDataEncryptorBuilder.

Commit:
Extract common codes from PGPPublicKeyEncryptedData and PGPSymmetricKeyEncryptedData.

Commit:
Add tests for PGPUserAttributeSubpacketVector, PGPCompressedData, PGPMarker, PGPOnePassSignature, PGPPadding, PGPSignature and PGPLiteralDataGenerator.

Commit:
Add key converter support for X448 in BcPGPKeyConverter and JcaPGPKeyConverter. Add tests for JcaPGPDigestCalculatorProviderBuilder, BcPGPKeyConverter, JcaPGPKeyConverter, and JcaPGPDigestCalculatorProviderBuilder.

Commit:
delete PQC namedGroups from hasAnyECDSA

Commit:
add kem namedGroup test

Commit:
Add tests for JcePBESecretKeyDecryptorBuilder, PGPSignatureVerifierBuilder, and PGPKeyRingGenerator

Commit:
Add tests for PGPSignatureVerifierBuilder.

Commit:
Add tests for ImageAttribute, PGPSignatureVerifierBuilder, PGPLiteralData, PGPSignature, JcaPGPContentSignerBuilder, JcaPGPKeyPair, PGPSignatureList and PGPEncryptedDataGenerator.

Commit:
Add tests for JcaPGPPublicKeyRing, PGPUserAttributeSubpacketVector, PGPLiteralData, and PGPUserAttributeSubpacketVectorGenerator.

Commit:
Add tests for JcaKeyFingerprintCalculator etc.

mwcw committed:
added exclusion to skip tests for 11,17 and 21 when testing for java 8.

Commit:
Add X25519, X448, Ed25519 and Ed448 tags for JcaPGPKeyConverter and BcPGPKeyConverter.

Commit:
Add X25519, X448, Ed25519 and Ed448 tags for BcPublicKeyDataDecryptorFactory, BcPublicKeyKeyEncryptionMethodGenerator, JcaPGPContentVerifierBuilderProvider, etc.

Commit:
Add tests for X25519, X448, Ed25519 and Ed448 tags.

Commit:
Refactor for X25519, X448, Ed25519 and Ed448 tags.

Commit:
Refactor for X25519, X448, Ed25519 and Ed448 tags.

Commit:
added exclusion to skip tests for 11,17 and 21 when testing for java 8.
Add tests for PGPOnePassSignatureList, JcePBESecretKeyDecryptorBuilder, PGPCompressedDataGenerator, and JcaPGPContentVerifierBuilderProvider.

Commit:
Update test_8.sh, removed extra space in test11 name

Commit:
Add tests for PGPCanonicalizedDataGenerator, JcePublicKeyDataDecryptorFactoryBuilder, and JcaPGPDigestCalculatorProviderBuilder.

Commit:
Merge remote-tracking branch 'origin/pg-jacoco' into pg-jacoco

Commit:
Add tests. Refactor on BcPGPKeyConverter and make functions in SExpression package-private

Commit:
Add tests to RFC6637Utils

Commit:
Merge branch 'main' into pg-jacoco

Commit:
Add tests to RFC6637Utils

Commit:
Add support for Camellia wrap in RFC6637Utils. Add tests for Ed25519 and Ed448 in BcImplProvider, for JcePublicKeyDataDecryptorFactoryBuilder, and Camellia wrap.

Commit:
Add support for Camellia AEAD cipher

Commit:
Add support for Camellia AEAD cipher

Commit:
Fix test failure in BcpgGeneralTest.

Commit:
Refactor around Camellia in AEAD situations.

peterdettman committed:
Cleanup obsolete TODO

Commit:
Add tests for CAMELLIA

Commit:
set default initial value for camellia

Commit:
rollback the setting for default iv of camellia

Commit:
rollback the setting for default iv of camellia

Commit:
Correct words in IllegalArgumentException. TODO: Derive Ed25519 and Ed448 from EdDsa

Commit:
Ed448PublicBCPGKey works now.

Commit:
Ed25519 and Ed448 works now.

Commit:
Pass all tests for Ed25519, Ed448, X25519 and X448.

Commit:
Add tests for curves

Commit:
Update SNTRUPrime KEMSpi for jdk21 (only tested generic)

Commit:
Fixed SNTRUPrime KEMSpi for jdk21, added more tests

mwcw committed:
source and javadoc generation

Commit:
Try to correct the encryption procedures of X25519 and X448 encryption.

Commit:
Correct the PKESK format of X25519 and X448.

Commit:
Jcajce classes uses HKDF for X25519 and X448 now

peterdettman committed:
Factor out TlsRsaKeyExchange class into core

Commit:
HKDF for X25519 and X448

Commit:
Refactor codes around X25519 and X448

Commit:
Refactor codes around X25519 and X448

Commit:
TODO task: unable to parse the test vector for Ed25519 key as the decrypted message is not correct.

Commit:
Remove some unused codes

Commit:
Changes around HKDF of X25519 and X448.

Commit:
Correct HKDF settings for X25519 and X448.

Commit:
correct info of for HKDF in BcPublicKeyDataDecryptorFactory and BcPublicKeyKeyEncryptionMethodGenerator

Commit:
Merge branch 'main' into pg-jacoco

peterdettman committed:
Inline RSA, PKCS1 code into TlsRsaKeyExchange

Commit:
Update the test case for X25519 with HKDF

Commit:
Use UserKeyingMaterialSpecWithPrepend for transporting HKDFParameters setting.

dghgit committed:
fixed OSGI export

dghgit committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

dghgit committed:
fixed OSGI export

dghgit committed:
fixed international characters in HTML

dghgit committed:
fixed OSGI export

dghgit committed:
fixed OSGI export

dghgit committed:
fixed OSGI export

dghgit committed:
fixed OSGI export

dghgit committed:
added debug stanza for subprojects

dghgit committed:
added OSGI/module-info to MLS

dghgit committed:
temp rm of mls

dghgit committed:
fix to mls build

dghgit committed:
removed provider depency from AllTests - not required

dghgit committed:
added use of JAVA_VERSION_PREFIX environment variable

Commit:
Update .gitlab-ci.yml file
Add artifact for mls test results.

dghgit committed:
Merge remote-tracking branch 'origin/pg-jacoco'

dghgit committed:
removed use of UserKeyingMaterialSpecWithPrepend.
minor refactoring to reduce public methods.
readded time converters for backwards compatibility (now call Utils).

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

peterdettman committed:
Refactor RSA engines

peterdettman committed:
Fix timing dependency on RSA output

mwcw committed:
Added config stanza for test report output dir. Updated job dependency for CI.

Commit:
pr-1567: Fix bug in createCompressed()

peterdettman committed:
Move some OID mappings around

mwcw committed:
Junit test results, report

mwcw committed:
Merge remote-tracking branch 'origin/main'

mwcw committed:
Gitlab test results for all test tasks.

dghgit committed:
refactored use of new ErrorBundle to avoid class loader issues with modules.

dghgit committed:
refactored to deal with class loader issues in ErrorBundles for modules

dghgit committed:
refactored to deal with class loader issues in ErrorBundles for modules

Commit:
Merge branch 'bcgit:main' into main

Honzaik committed:
update to draft v13. samples from draft are now compatible with the exception that the Dilithium private key seems to have a different encoding. separated samples from CertTest into resource files.

brownp2k committed:
Add targets for build-util and build-pkix.

Commit:
reverted otherInfo to null
fixed exceptions
added testKEMVec() uses sntrup653 count 0 for testing no kdf (ss & ct)

Commit:
excluded broken mls test (waiting for fixed tests)

dghgit committed:
fixed unknown type error

dghgit committed:
added missing cast...

dghgit committed:
move of oiw, cryptlib and mozilla to util package

dghgit committed:
move of ASN.1 gnu, ntt and iana to util package

dghgit committed:
move of ASN.1 isara, iso, kisa, and microsoft to util package

peterdettman committed:
Refactoring

peterdettman committed:
Fix headings

peterdettman committed:
Belated release note for BCJSSE change

dghgit committed:
move of ASN.1 edec, misc, nsri, and rosstandart  to util package

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

peterdettman committed:
Add ASN1RelativeOID cache

dghgit committed:
correct suppressions.xml for Windows

dghgit committed:
correct suppressions.xml for Windows/Linux

dghgit committed:
moved to platform independent path seperator

dghgit committed:
moved to platform independent path seperator

dghgit committed:
further windows tweaks

peterdettman committed:
ASN.1: Limit OID contents to 4096 bytes

peterdettman committed:
Use tryFromID to check for OID string

cipherboy committed:
Fix for EdDSA verification infinite loop

Resolves: https://github.com/bcgit/bc-java/issues/1599

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

cipherboy committed:
Add regression tests for Ed25519 loop

This adds regression tests for the Ed25519 loop bug, generated using an
instrumented build. Vectors were found by adapting affected functions
(addShifted_NP, subShifted_NP, and reduceBasisVar) to account for
different parameters (last and s in the first two or number of loop
iterations in the latter), to ensure correct behavior and sufficient
coverage. Vectors were then cross-verified with an alternative
implementation, though signature generation was not impacted, only
verification.

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

Commit:
enabled broken mls tests (now fixed)

dghgit committed:
fixed typo in releasenotes.html
tweaking of exception throwing in CMP to improve exception messages.

dghgit committed:
Merge branch 'stellar/r1rv77-buildFixes' of https://github.com/stellarscience/bc-java into stellarscience-stellar/r1rv77-buildFixes

dghgit committed:
Merge branch 'stellarscience-stellar/r1rv77-buildFixes'

dghgit committed:
added sanity checks for boolean flags - related to github #1575

dghgit committed:
conflict resolution

dghgit committed:
conflict resolution

dghgit committed:
Merge branch 'Honzaik-main'

dghgit committed:
standard digest tests for PGP (draft) relates to GitHub #1566

dghgit committed:
added use of TestResourceFinder.

dghgit committed:
Initial cut of RFC 9215 for GOST-3410 relates to github #1586

dghgit committed:
additional work on RFC 9215 parameter restrictions github #1586

dghgit committed:
additional work on RFC 9215 parameter restrictions, more nulls - github #1586

dghgit committed:
additional work on RFC 9215 parameter restrictions, capturing reduced sequence size - github #1586

dghgit committed:
added setting of digestParam for old GOST params

dghgit committed:
minor cleanup

dghgit committed:
removed external dependency on provider composite name class
standardised names to RFC draft.

dghgit committed:
256 bit NTRU parameters

dghgit committed:
added support for NTRU HRSS 1373

dghgit committed:
minor refactoring - removal of common code.

dghgit committed:
added additional NTRU OIDs to provider.

dghgit committed:
change to use this.read() relates to github #1577

dghgit committed:
Merge branch 'memoable' of https://github.com/seungyeonpark/bc-java into seungyeonpark-memoable

dghgit committed:
Merge branch 'seungyeonpark-memoable'

peterdettman committed:
Further checks in bcpg.sig

peterdettman committed:
EdDSA: Explicit guard against infinite looping

peterdettman committed:
Remove duplicate test case

dghgit committed:
fixed dodgy test...

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
Initial cut for id-alg-cek-hkdf-sha256 CMS support.

dghgit committed:
added import of java.logging (github #1608)

dghgit committed:
added new RSA mode for better TLS unwrap operation relates to github #1528

dghgit committed:
widened RSA private key support - relates to github #1528

dghgit committed:
corrected repackaging issue.

dghgit committed:
added missing copy statement

dghgit committed:
removed AllTests for keybox test

mwcw committed:
Added running ant build to CI.

mwcw committed:
Added ant to path.

mwcw committed:
Update to pass multiple args to javadoc

Commit:
Added call to gitlab_wait

Commit:
Removed gitlab_wait.sh from jobs.

dghgit committed:
added some explanatory comments.

dghgit committed:
minor changes to messages

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

peterdettman committed:
Round out implementation of TLS RSA PreMasterSecret

peterdettman committed:
Refactoring

dghgit committed:
removed use of PQC provider.

Commit:
NTRU KEMspi for jdk 21

Commit:
Merge branch 'tls-kyber' of https://github.com/samsungsds-research-papers/bc-java into samsungsds-research-papers-tls-kyber

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
digest name fix.
added jdk1.5 compatibility classes.

dghgit committed:
fixed checkstyle issue

dghgit committed:
reverted PGPUtil inclusion

peterdettman committed:
Fix various bitstring usages

peterdettman committed:
Fix engineGetOutputSize and add some TODOs

cipherboy committed:
Add regression tests for Ed448 loop

See also: 02f1fb441b6b72826ea892aea4a741403d9562f3

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

peterdettman committed:
Update EdDSA tests

dghgit committed:
corrected test to avoid failure where last byte was already zero.

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

Commit:
Added Mock Tls Tests

Commit:
Renamed kyber to mlkem

Commit:
Added mismatch test for different version of mlkem

Commit:
Added missing input validation for ML-KEM Encapsulation (fips 203 6.2)

Commit:
Replaced RuntimeException with IllegalArgumentException

dghgit committed:
minor changes for compatibility

dghgit committed:
minor changes for compatibility

dghgit committed:
added AES CBC based PKCS12 store and AES GCM based PKCS12 store

dghgit committed:
minor changes for compatibility

dghgit committed:
added use of Strings.toLowerCase()

peterdettman committed:
BCJSSE: Private key may not support encoding

peterdettman committed:
BCJSSE: Improved workaround for InetAddress limitation

- URLConnectionUtil now calls BCSSLSocket.setHost instead of direct SNI config

Commit:
Added KEM check in selectKeyShare/Group
minor name change

Commit:
added fail in catch

Commit:
quick little arrangement

peterdettman committed:
TLS: Rework ML-KEM PR

Commit:
Merge remote-tracking branch 'origin/samsungsds-research-papers-tls-kyber'

# Conflicts:
#	tls/src/main/java/org/bouncycastle/jsse/provider/NamedGroupInfo.java
#	tls/src/main/java/org/bouncycastle/tls/NamedGroup.java
#	tls/src/main/java/org/bouncycastle/tls/TlsServerProtocol.java
#	tls/src/main/java/org/bouncycastle/tls/TlsUtils.java
#	tls/src/main/java/org/bouncycastle/tls/crypto/TlsCrypto.java
#	tls/src/main/java/org/bouncycastle/tls/crypto/impl/bc/BcTlsCrypto.java
#	tls/src/main/java/org/bouncycastle/tls/crypto/impl/jcajce/JcaTlsCrypto.java

Commit:
Merged tests to tls ml-kem

Commit:
removed extra files due to naming

Commit:
removed extra files due to naming

dghgit committed:
added role back to simpler match for certificate matching (issue around locating trust anchors).

dghgit committed:
minor formatting

dghgit committed:
minor formatting

dghgit committed:
moved to 1.78, BCJSSE 1.0.19.

dghgit committed:
minor reformatting

dghgit committed:
compatibility updates.

mwcw committed:
Ensure test data changes are pulled into container.

mwcw committed:
Removed test data pull.

peterdettman committed:
Cleanup deprecated TLS code

peterdettman committed:
DTLS: more robust exclusion of stream ciphers

peterdettman committed:
TLS: Use more appropriate method to check for ECDH curve

dghgit committed:
Java 5 compatibility updates

dghgit committed:
Java 5 compatibility updates

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
Java 5 compatibility updates.

dghgit committed:
Java 5 compatibility updates.

dghgit committed:
Java 5 compatibility updates.

cipherboy committed:
Add release notes for 1.78

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

cipherboy committed:
Fix CCM input length check

As reported by @CipherGato, CCM fails to decrypt a full-length message
for n=13, q=2 (maximum length 65,535). This is because the input message
already includes the flags and nonce components (B0 from SP800 38C).
Correctly handle full-length CCM decryption.

Resolves: #1570

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

dghgit committed:
Java 5 compatibility

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
Java 4 compatibility updates.
Removed Camellia wrap enable property from OpenPGP (not required).

mtgag committed:
added name resolving for jurisdiction{C,ST,L}. These are used in EV certificates.

cipherboy committed:
Add CCM fix to release notes

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

dghgit committed:
Merge branch 'mtgag-jurisdiction_strings'

dghgit committed:
Java 4 updates

peterdettman committed:
Update release notes

dghgit committed:
Java 4 updates

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
minor cleanups

dghgit committed:
moved to 1.78

dghgit committed:
corrected version number

dghgit committed:
almost final releasenotes

dghgit committed:
update

dghgit committed:
Java 2, Java 3, compatibility updates

sergejskozlovics committed:
Merge remote-tracking branch 'upstream/main' into tls-injection2

sergejskozlovics committed:
moved the main TLS Injection Mechanism out from the core package, leaving only Asn1Bridge and Asn1BridgeForInjectedSigAlgs there; no dependency on the tls package

sergejskozlovics committed:
refactored TLS Injection Mechanism; moved it to the tls package

sergejskozlovics committed:
fixed comments

sergejskozlovics committed:
Removed UniversalKeyPairGenerator, since it is not used by the TLS Injection Mechanism. Besides, UniversalKeyPairGenerator is difficult to implement.

sergejskozlovics committed:
renamed variables "*EncodedKey" to "*ByteKey"; add "RSASSA-PSS" as a fallback algorithm for the BC adapter (since <hash-name>WITHRSAANDMGF1 sometimes is not found)

sergejskozlovics committed:
added .clone() for the shared key, since BC may alter it

sergejskozlovics committed:
added isEmpty() to InjectionPoint

sergejskozlovics committed:
fixed a bug when KEM keyGen() was unnecessarily invoked at the server side

sergejskozlovics committed:
adding the BC provider in InjectionPoint.push
